src/System.Private.ServiceModel/tests/Scenarios/Contract/Message/MessageContractTests.cs

```diff
@@ -119,4 +119,60 @@ public static void MessageBody_Elements_CustomerElement_Value_Matches()
 
         Assert.True(errorBuilder.Length == 0, errorBuilder.ToString());
     }
-}
\ No newline at end of file
+
+    [Fact]
+    [OuterLoop]
+    public static void MessageHeader_MustUnderstand_True()
+    {
+        StringBuilder errorBuilder = new StringBuilder();
+
+        try
+        {
+            var headers = MessageContractHelpers.SetupMessageContractTestsAndGetHeaders("MessageContractRequestReplyWithMessageHeader");
+            int index = headers.FindHeader(MessageContractConstants.extraValuesName, MessageContractConstants.extraValuesNamespace);
+            var header = headers[index];
+            if (header == null)
+            {
+                errorBuilder.AppendLine("There's no header in the message.");
+            }
+            else if (!header.MustUnderstand)
+            {
+                errorBuilder.AppendLine("Expected MustUnderstand to be true, but it was false.");
+            }
+        }
+        catch (Exception ex)
+        {
+            errorBuilder.AppendLine(String.Format("Unexpected exception was caught: {0}", ex.ToString()));
+        }
+
+        Assert.True(errorBuilder.Length == 0, errorBuilder.ToString());
+    }
+
+    [Fact]
+    [OuterLoop]
+    public static void MessageHeader_MustUnderstand_False()
+    {
+        StringBuilder errorBuilder = new StringBuilder();
+
+        try
+        {
+            var headers = MessageContractHelpers.SetupMessageContractTestsAndGetHeaders("MessageContractRequestReplyWithMessageHeaderNotNecessaryUnderstood");
+            int index = headers.FindHeader(MessageContractConstants.extraValuesName, MessageContractConstants.extraValuesNamespace);
```

Comment:
Nit: prefer we don't use 'var' unless it's really obvious from contact what type is being returned.